Source/MLX/MLXArray+Indexing.swift

```diff
@@ -521,4 +353,1209 @@ extension MLXArray {
+        return (
+            [],
+            update.broadcast(to: shape32, stream: stream),
+            []
+        )
+    }
+
+    // Analyse the types of the indices
+    var maxDimensions = 0
+    var arraysFirst = false
+    var countNewAxis = 0
+    var countSlices = 0
+    var countArrays = 0
+    var countStridedSlices = 0
+    var countSimpleSlicesPost = 0
+
+    var haveArray = false
+    var haveNonArray = false
+
+    for item in operations {
+        switch item {
+        case .ellipsis:
+            // has been converted to slices
+            fatalError("unexpected item: \(item) in scatterArguments")
+
+        case .newAxis:
+            haveNonArray = true
+            countNewAxis += 1
+
+        case .index:
+            // ignore
+            break
+
+        case .slice(let slice):
+            haveNonArray = haveArray
+            countSlices += 1
+            if slice.stride == 1 {
+                countSimpleSlicesPost += 1
+            } else {
+                countStridedSlices += 1
+            }
+
+        case .array(let array):
+            haveArray = true
+            if haveArray && haveNonArray {
+                arraysFirst = true
+            }
+            maxDimensions = max(array.ndim, maxDimensions)
+            countArrays += 1
+            countSimpleSlicesPost = 0
+        }
+    }
+
+    // We have index dims for the arrays, strided slices (implemented as arrays), none
+    var indexDimensions = maxDimensions + countNewAxis + countSlices - countSimpleSlicesPost
+
+    // If we have simple non-strided slices, we also attach an index for that
+    if indexDimensions == 0 {
+        indexDimensions = 1
+    }
+
+    // Go over each index type and translate to the needed scatter args
+    var arrayIndices = [MLXArray]()
+    var sliceNumber = 0
+    var arrayNumber = 0
+    var axis = 0
+
+    // We collect the shapes of the slices and updates during this process
+    var updateShape = Array(repeating: 1, count: nonNewAxisOperationCount)
+    var sliceShapes = [Int]()
+
+    for item in operations {
+        switch item {
+        case .ellipsis:
+            // has been converted to slices
+            fatalError("unexpected item: \(item) in scatterArguments")
+
+        case .newAxis:
+            // We only use the newAxis's for bookeeping dimensions
+            sliceNumber += 1
+
+        case .index(let index):
+            arrayIndices.append(src.resolve(index: index, axis: axis))
+            updateShape[axis] = 1
+            axis += 1
+
+        case .slice(let slice):
+            let size = src.dim(axis).int32
+            let start = slice.absoluteStart(size)
+            let end = slice.absoluteEnd(size)
+            let stride = slice.stride
+
+            var indexShape = Array(repeating: 1, count: indexDimensions)
+
+            // If it's a simple slice, we only need to add the start index
+            if arrayNumber >= countArrays && stride == 1 {
```

Comment:
See also https://github.com/ml-explore/mlx/issues/937

Originally the c++ code was:

```
if (array_num >= num_arrays && num_strided_slices <= 0 && stride == 1) {
```

This seems to work, but verify with the fix we actually take for that bug.